File: tests/prototypes/controller.lua
---@alias Action 'show'|'hide'|'toggle'
---@alias Observation 'absent'|'hidden'|'here'|'elsewhere'|'ambiguous'
---@alias Progress 'idle'|'uncertain_show'|'uncertain_hide'|'confirmed_show'|'confirmed_hide'
---@class PrototypeRequest
---@field action Action
---@field observed Observation
---@field progress Progress
---@field owned boolean

local function parse(args)
  if #args ~= 4 then return nil, 'invalid_request' end
  local sets = {
    { show = true, hide = true, toggle = true },
    { absent = true, hidden = true, here = true, elsewhere = true, ambiguous = true },
    { idle = true, uncertain_show = true, uncertain_hide = true, confirmed_show = true, confirmed_hide = true },
    { verified = true, unverified = true },
  }
  local errors = { 'invalid_action', 'invalid_observation', 'invalid_progress', 'invalid_ownership' }
  for i, allowed in ipairs(sets) do
    if not allowed[args[i]] then return nil, errors[i] end
  end
  return { action = args[1], observed = args[2], progress = args[3], owned = args[4] == 'verified' }
end

---@param r PrototypeRequest
local function plan(r)
  if r.progress == 'uncertain_show' or r.progress == 'uncertain_hide' then
    return 'observe', 'uncertain', true
  end
  if r.observed == 'ambiguous' or (not r.owned and r.observed ~= 'absent') then
    return 'none', 'needs_attention', true
  end
  if r.progress == 'confirmed_show' then return 'none', 'show', false end
  if r.progress == 'confirmed_hide' then return 'none', 'hide', false end
  local show = r.action == 'show' or (r.action == 'toggle' and r.observed ~= 'here')
  if show then
    if r.observed == 'absent' then return 'spawn', 'show', true end
    if r.observed == 'here' then return 'none', 'show', false end
    return 'move', 'show', true
  end
  if r.observed == 'here' or r.observed == 'elsewhere' then return 'park', 'hide', true end
  return 'none', 'hide', false
end

local request, err = parse(arg)
if err then
  io.stdout:write(vim.json.encode({ error = err }) .. '\n')
  vim.cmd('cquit 2')
else
  local effect, outcome, retain_guard = plan(request)
  io.stdout:write(vim.json.encode({ effect = effect, outcome = outcome, retain_guard = retain_guard }) .. '\n')
  vim.cmd('qa!')
end

File: tests/prototypes/controller.rs
use std::env;

#[derive(Clone, Copy)]
enum Action {
    Show,
    Hide,
    Toggle,
}
#[derive(Clone, Copy)]
enum Observed {
    Absent,
    Hidden,
    Here,
    Elsewhere,
    Ambiguous,
}
#[derive(Clone, Copy)]
enum Progress {
    Idle,
    UncertainShow,
    UncertainHide,
    ConfirmedShow,
    ConfirmedHide,
}

struct Request {
    action: Action,
    observed: Observed,
    progress: Progress,
    owned: bool,
}

fn parse(args: &[String]) -> Result<Request, &'static str> {
    if args.len() != 4 {
        return Err("invalid_request");
    }
    let action = match args[0].as_str() {
        "show" => Action::Show,
        "hide" => Action::Hide,
        "toggle" => Action::Toggle,
        _ => return Err("invalid_action"),
    };
    let observed = match args[1].as_str() {
        "absent" => Observed::Absent,
        "hidden" => Observed::Hidden,
        "here" => Observed::Here,
        "elsewhere" => Observed::Elsewhere,
        "ambiguous" => Observed::Ambiguous,
        _ => return Err("invalid_observation"),
    };
    let progress = match args[2].as_str() {
        "idle" => Progress::Idle,
        "uncertain_show" => Progress::UncertainShow,
        "uncertain_hide" => Progress::UncertainHide,
        "confirmed_show" => Progress::ConfirmedShow,
        "confirmed_hide" => Progress::ConfirmedHide,
        _ => return Err("invalid_progress"),
    };
    let owned = match args[3].as_str() {
        "verified" => true,
        "unverified" => false,
        _ => return Err("invalid_ownership"),
    };
    Ok(Request {
        action,
        observed,
        progress,
        owned,
    })
}

fn plan(r: Request) -> (&'static str, &'static str, bool) {
    if matches!(
        r.progress,
        Progress::UncertainShow | Progress::UncertainHide
    ) {
        return ("observe", "uncertain", true);
    }
    if matches!(r.observed, Observed::Ambiguous)
        || (!r.owned && !matches!(r.observed, Observed::Absent))
    {
        return ("none", "needs_attention", true);
    }
    match r.progress {
        Progress::ConfirmedShow => return ("none", "show", false),
        Progress::ConfirmedHide => return ("none", "hide", false),
        _ => (),
    }
    let show = match r.action {
        Action::Show => true,
        Action::Hide => false,
        Action::Toggle => !matches!(r.observed, Observed::Here),
    };
    match (show, r.observed) {
        (true, Observed::Absent) => ("spawn", "show", true),
        (true, Observed::Hidden | Observed::Elsewhere) => ("move", "show", true),
        (true, Observed::Here) => ("none", "show", false),
        (false, Observed::Here | Observed::Elsewhere) => ("park", "hide", true),
        (false, _) => ("none", "hide", false),
        _ => unreachable!(),
    }
}

fn main() {
    match parse(&env::args().skip(1).collect::<Vec<_>>()) {
        Ok(r) => {
            let (effect, outcome, retain_guard) = plan(r);
            println!("{{\"effect\":\"{effect}\",\"outcome\":\"{outcome}\",\"retain_guard\":{retain_guard}}}");
        }
        Err(code) => {
            println!("{{\"error\":\"{code}\"}}");
            std::process::exit(2);
        }
    }
}
